### ISSUE TITLE
Inconsistent bounds checking behavior on *p++ where p is an array_ptr of array_ptr or nt_array_ptr

### ISSUE THREAD

**Issue by dopelsunce** [microsoft/checkedc-clang#880]
Currently, the following code compiles and runs without error:
```
void foo() {
    char s _Nt_checked[] = "abc";
    _Array_ptr<char> arr2 _Checked[3] = {};
    _Array_ptr<_Array_ptr<char>> p : bounds(arr2, arr2 + 3) = arr2;
    *p++ = s;
}
```

Note `p` does not use self-referential bounds so `*p++` does not trigger the undefined behavior in #863.

However, changing the type of `p` to ` an array of_Nt_array_ptr<char>` gives a compiler error. That is,
```
void foo() {
    char s _Nt_checked[] = "abc";
    _Nt_array_ptr<char> arr2 _Checked[3] = {};
    _Array_ptr<_Nt_array_ptr<char>> p : bounds(arr2, arr2 + 3) = arr2;
    *p++ = s;
}
```
Outputs
```
test.c:5:5: error: inferred bounds 'bounds(*p++, *p++ + 0)' contain a modifying expression; use a temporary instead
    *p++ = s;
    ^~~~
test.c:5:6: note: modifying expression
    *p++ = s;
     ^~~
1 error generated.
```

First, I expect both programs to have the same compile-time output, regardless of the inner array type. `*p++` operates on the outer array, so the type of the inner array should not matter. The compiler may either accept both or reject both, whichever complies to the spec (I'm not sure if this is defined).

Second, the inferred bounds `bounds(*p++, *p++ + 0)` in the error message of the second program look weird. The bounds should concern `p` rather than `*p`. Maybe it should be `(p++, p++ + 0)` instead?

**Comment by kkjeer** [comment]
Currently, we don't know the target bounds of a pointer stored in a pointer dereference, unless it is a `_Ptr` type or an `_Nt_array_ptr` (see the method `UnaryOperatorTargetBounds` in SemaBounds.cpp).

In the first function, the dereference expression `*p++` has type `_Array_ptr<char>`, since `p` has type `_Array_ptr<_Array_ptr<char>>`. Since `_Array_ptr<char>` is not a `_Ptr` or `_Nt_array_ptr` type, the target bounds of `*p++` are `bounds(unknown)`. These bounds do not contain a modifying expression.

In the second function, the dereference expression `*p++` has type `_Nt_array_ptr<char>`, since `p` has type `_Array_ptr<_Nt_array_ptr<char>>`. Since `_Nt_array_ptr<char>` is an `_Nt_array_ptr` type, the target bounds of `*p++` are `bounds(*p++, *p++ + 0)` (as computed by the method `CreateTypeBasedBounds` in SemaBounds.cpp). These bounds do contain a modifying expression `*p++`, which fails the `S.CheckNonModifyingBounds` check in the `CheckBinaryOperator` method.
